src/main/java/org/opensearch/ad/util/BulkUtil.java

```diff
@@ -37,25 +37,73 @@
 import org.opensearch.action.bulk.BulkItemResponse;
 import org.opensearch.action.bulk.BulkRequest;
 import org.opensearch.action.bulk.BulkResponse;
+import org.opensearch.action.index.IndexRequest;
 
 public class BulkUtil {
     private static final Logger logger = LogManager.getLogger(BulkUtil.class);
 
-    public static List<DocWriteRequest<?>> getIndexRequestToRetry(BulkRequest bulkRequest, BulkResponse bulkResponse) {
-        List<DocWriteRequest<?>> res = new ArrayList<>();
+    public static List<IndexRequest> getIndexRequestToRetry(BulkRequest bulkRequest, BulkResponse bulkResponse) {
+        List<IndexRequest> res = new ArrayList<>();
 
         Set<String> failedId = new HashSet<>();
         for (BulkItemResponse response : bulkResponse.getItems()) {
-            if (response.isFailed()) {
+            if (response.isFailed() && ExceptionUtil.isRetryAble(response.getFailure().getStatus())) {
                 failedId.add(response.getId());
             }
         }
 
         for (DocWriteRequest<?> request : bulkRequest.requests()) {
-            if (failedId.contains(request.id())) {
-                res.add(request);
+            try {
+                if (failedId.contains(request.id())) {
+                    res.add(cloneIndexRequest((IndexRequest) request));
+                }
+            } catch (ClassCastException e) {
+                logger.error("We only support IndexRequest");
```

Comment:
Can this log include the type of IndexRequest to give more insight into failure? Or will this be propagated with the thrown error?

Reply:
yes, added exception stack trace in the log.